File: lab-2/C#/task-3/Fraction.cs
namespace Task3
{
    public class Fraction
    {
        public int Numerator { get; set; }

        private int _denominator = 1;
        public int Denominator
        {
            get => _denominator;
            set
            {
                if (value == 0)
                {
                    throw new ArgumentException("Знаменник не може бути нулем");
                }
                if (value < 0)
                {
                    Numerator = -Numerator;
                    _denominator = -value;
                }
                else
                {
                    _denominator = value;
                }
            }
        }

        public Fraction()
            : this(0) { }

        public Fraction(int numerator)
            : this(numerator, 1) { }

        public Fraction(int numerator, int denominator)
        {
            Numerator = numerator;
            Denominator = denominator;
        }

        private static int GCD(int a, int b)
        {
            a = Math.Abs(a);
            b = Math.Abs(b);
            while (b != 0)
            {
                int temp = b;
                b = a % b;
                a = temp;
            }
            return a;
        }

        public Fraction Reduce()
        {
            int gcd = GCD(Numerator, _denominator);
            Numerator /= gcd;
            _denominator /= gcd;
            return this;
        }

        public Fraction Add(Fraction other)
        {
            Numerator = Numerator * other._denominator + other.Numerator * _denominator;
            _denominator *= other._denominator;
            return Reduce();
        }

        public Fraction Add(int number) => Add(new Fraction(number));

        public Fraction Subtract(Fraction other)
        {
            Numerator = Numerator * other._denominator - other.Numerator * _denominator;
            _denominator *= other._denominator;
            return Reduce();
        }

        public Fraction Multiply(Fraction other)
        {
            Numerator *= other.Numerator;
            _denominator *= other._denominator;
            return Reduce();
        }

        public Fraction Multiply(int number) => Multiply(new Fraction(number));

        public Fraction Divide(Fraction other)
        {
            if (other.Numerator == 0) throw new DivideByZeroException();

            Numerator *= other._denominator;
            _denominator *= other.Numerator;
            
            return Reduce();
        }

        public bool Equals(Fraction other)
        {
            Fraction f1 = new Fraction(Numerator, _denominator).Reduce();
            Fraction f2 = new Fraction(other.Numerator, other._denominator).Reduce();
            return f1.Numerator == f2.Numerator && f1._denominator == f2._denominator;
        }

        public override string ToString() => $"{Numerator}/{_denominator}";
    }
}

File: lab-2/C++/task-3/Fraction.cpp
#include "./Fraction.hpp"

Fraction::Fraction() : Fraction(0) {}

Fraction::Fraction(int numerator) : Fraction(numerator, 1) {}

Fraction::Fraction(int numerator, int denominator)
{
  this->numerator = numerator;
  setDenominator(denominator);
}

Fraction &Fraction::reduce()
{
  int gcd = _gcd(abs(numerator), abs(denominator));

  numerator /= gcd;
  denominator /= gcd;

  return *this;
}

Fraction &Fraction::add(const Fraction &other)
{
  numerator = numerator * other.denominator + other.numerator * denominator;
  denominator *= other.denominator;
  
  return reduce();
}

Fraction &Fraction::add(int number)
{
  return add(*new Fraction(number));
}

Fraction &Fraction::subtract(const Fraction &other)
{
  numerator = numerator * other.denominator - other.numerator * denominator;
  denominator *= other.denominator;
  
  return reduce();
}

Fraction &Fraction::multiply(const Fraction &other)
{
  numerator *= other.numerator;
  denominator *= other.denominator;

  return reduce();
}

Fraction &Fraction::multiply(int number)
{
  return multiply(*new Fraction(number));
}

Fraction &Fraction::divide(const Fraction &other)
{
  if (other.numerator == 0) throw runtime_error("Attempted to divide by Zero\n");

  numerator *= other.denominator;
  denominator *= other.numerator;
  
  return reduce();        
}

bool Fraction::equals(const Fraction &other)
{
    Fraction f1 = (new Fraction(numerator, denominator))->reduce();
    Fraction f2 = (new Fraction(other.numerator, other.denominator))->reduce();

    return f1.numerator == f2.numerator && f1.denominator == f2.denominator;
}

void Fraction::print() const
{
  if (denominator > 0)
  {
    cout << numerator << "/" << denominator << endl;
  }
  else
  {
    cout << numerator << endl;
  }
}